test/gtest/common/test_helpers.cc

```diff
@@ -15,6 +15,133 @@ namespace ucs {
 
 const double test_timeout_in_sec = 60.;
 
+const double watchdog_timeout_default = 900.; // 15 minutes
+
+static test_watchdog_t watchdog;
+
+void *watchdog_func(void *arg)
+{
+    int ret = 0;
+    double now;
+    struct timespec timeout;
+
+    pthread_mutex_lock(&watchdog.mutex);
+
+    /* sync with parent thread */
+    pthread_barrier_wait(&watchdog.barrier);
+
+    do {
+        now = ucs_get_accurate_time();
+        ucs_sec_to_timespec(now + watchdog.timeout, &timeout);
+
+        ret = pthread_cond_timedwait(&watchdog.cv, &watchdog.mutex, &timeout);
+        if (!ret) {
+            switch (watchdog.state) {
+            case test_watchdog_t::WATCHDOG_RUN:
+                if (watchdog.timeout != watchdog_timeout_default) {
+                    /* reset to the default one */
+                    watchdog.timeout = watchdog_timeout_default;
+                }
+                break;
+            case test_watchdog_t::WATCHDOG_STOP:
+                /* force the end of the cycle */
+                ret = 1;
+                break;
+            case test_watchdog_t::WATCHDOG_TIMEOUT_SET:
+                /* this value will be reseted to the default one when signal
+                 * will be received when the test completed */
+                watchdog.state = test_watchdog_t::WATCHDOG_RUN;
+                break;
+            }
+            pthread_barrier_wait(&watchdog.barrier);
+        }
+    } while (!ret);
+
+    if (ret && (watchdog.state != test_watchdog_t::WATCHDOG_STOP)) {
```

Comment:
ret can't be 0 here, right? so, no need to check it here

Reply:
good catch! fixed